### PR TITLE
Propagate metric metadata via remote write

### PR DESCRIPTION
**Pull request by gotjosh**
This PR takes a stab at that by using a similar approach to the WAL watcher. We observe the `scrapeCache` based on a specified period and pull the available metadata to send it to remote storage.

A high-level diagram of the process looks like:

mermaid-diagram-20200213152256

Documentation that explains the thinking behind this design lives in [this design doc](https://docs.google.com/document/d/1LoCWPAIIbGSq59NG3ZYyvkeNb8Ymz28PUKbg_yhAzvE/edit#heading=h.iudi25q1yiao). The [first version of the design doc](https://docs.google.com/document/d/1GoWgKuioZhjFcDNyyrs_WrKxhHR8UZNFCTqkpt04k4M/edit) contained details about other parts not relevant to Prometheus. Highly advise to only analyse the former.

Signed-off-by: gotjosh <josue@grafana.com>

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -665,6 +673,15 @@ type QueueConfig struct {
 	MaxBackoff model.Duration `yaml:"max_backoff,omitempty"`
 }
 
+// MetadataConfig is the configuration for sending metadata to remote
+// storage.
+type MetadataConfig struct {
+	// Send controls whenever we send metric metadata to remote storage.
+	Send bool `yaml:"send"`
+	// MetadataSendInterval controls how frequently we send metric metadata.
```
Comment (rfratto) [inline config/config.go]: nit: the comment wasn't updated when the struct member was changed

File: prompb/remote.proto [hunk 2]
```diff
@@ -21,6 +21,11 @@ import "gogoproto/gogo.proto";
 
 message WriteRequest {
   repeated prometheus.TimeSeries timeseries = 1 [(gogoproto.nullable) = false];
+  // Cortex uses this field to determine the source of the write request.
+  // We reserve it to avoid any compatibility issues.
+  reserved  2;
```
Comment (rfratto) [inline prompb/remote.proto]: smallest, tiniest nit: there's an extra space after `reserved` :)

File: storage/remote/queue_manager.go [hunk 3]
```diff
@@ -261,18 +301,23 @@ type QueueManager struct {
...
 	maxNumShards               prometheus.Gauge
 	minNumShards               prometheus.Gauge
 	desiredNumShards           prometheus.Gauge
 	bytesSent                  prometheus.Counter
+	metadataBytesSent          prometheus.Counter
 }
 
 // NewQueueManager builds a new QueueManager.
-func NewQueueManager(metrics *queueManagerMetrics, watcherMetrics *wal.WatcherMetrics, readerMetrics *wal.LiveReaderMetrics, logger log.Logger, walDir string, samplesIn *ewmaRate, cfg config.QueueConfig, externalLabels labels.Labels, relabelConfigs []*relabel.Config, client StorageClient, flushDeadline time.Duration) *QueueManager {
+func NewQueueManager(metrics *queueManagerMetrics, watcherMetrics *wal.WatcherMetrics, readerMetrics *wal.LiveReaderMetrics, logger log.Logger, walDir string, samplesIn *ewmaRate, mCfg config.MetadataConfig, cfg config.QueueConfig, externalLabels labels.Labels, relabelConfigs []*relabel.Config, client StorageClient, flushDeadline time.Duration, sm scrape.ReadyManager) *QueueManager {
```
Comment (rfratto) [inline storage/remote/queue_manager.go]: this 100% isn't your fault, but the number of parameters here makes me sad :slightly_frowning_face:

Comment (gotjosh) [reply storage/remote/queue_manager.go:320]: I tried to fight this one, but it was no good. I just gave in. 😭

Comment (rfratto) [reply storage/remote/queue_manager.go:320]: oh, yeah, I don't think anything we could do here will make it upstream, unfortunately

**Pull request merged**